src/utils/adjust-markup.js

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright 2019 Adobe. All rights reserved.
+ * This file is licensed to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+const { select } = require('hast-util-select');
+const fromDOM = require('hast-util-from-dom');
+const { JSDOM } = require('jsdom');
+const jquery = require('jquery')(new JSDOM().window);
+
+/**
+ * Load the markup config file.
+ */
+async function loadMarkupConfig() {
+// TODO: properly connect with markup.yaml loading logic
+//   return await new MarkupConfig()
+//     .withSource('markup.yaml')
+//     .init();
+  return {
+    markup: [
+      {
+        name: 'foo',
+        match: 'paragraph',
+        class: 'bar',
+        attribute: {
+          baz: 'qux',
+        },
+        wrap: '<div class="corge">',
+        type: 'markdown',
+      },
+      {
+        name: 'grault',
+        match: '.corge',
+        class: 'grault',
+        attribute: {
+          garply: 'waldo',
+        },
+        wrap: '<div class="fred">',
+      },
+    ],
+  };
+}
+
+/**
+ * Adjust the MDAST tree according to the markup config.
+ * This is done by registering new matchers on the VDOMTransformer before the HTML is generated
+ * in the pipeline
+ *
+ * @param {Object} context the execution context
+ * @param {Object} logger the pipeline logger
+ * @param {Object} transformer the VDOM transformer
+ */
+async function adjustMDAST(context, { logger, transformer }) {
+  const config = await loadMarkupConfig();
+  if (!config.markup) {
+    return;
+  }
+
+  logger.debug('Applying custom MDAST markup config');
+
+  config.markup
+    .filter((cfg) => cfg.type === 'markdown')
+    .forEach((cfg) => {
+      transformer.match(cfg.match, (h, node) => {
+        const handler = transformer.constructor.default(node);
+        const el = handler(h, node);
+
+        // Append classes to the element (space or comma separated)
+        if (cfg.class) {
+          el.properties.className = [...(el.properties.className || '').split(' '), ...cfg.class.split(/[ ,]/)].join(' ').trim();
+        }
+
+        // Append attributes to the element
+        if (cfg.attribute) {
+          Object.assign(el.properties, cfg.attribute);
+        }
+
+        // Wrap the element
+        if (cfg.wrap) {
+          const doc = new JSDOM().window.document;
```

Comment:
Not a big fan of the current approach… having to parse the HTML back to HAST to transform it back to HTML seems a waste of resources, but my goal here was to let the `markdown` config modify the HAST, while the `html` config modifies the DOM.